### PR TITLE
[13.0][MIG] hr_expense_advance_clearing

### PR DESCRIPTION
**Pull request by kittiu**
Standard migration, ready for review.

### CODE REVIEW

File: hr_expense_advance_clearing/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2019 Kitti Upariphutthiphong <kittiu@ecosoft.co.th>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Employee Advance and Clearing",
+    "version": "13.0.1.0.0",
+    "category": "Human Resources",
+    "author": "Ecosoft, " "Odoo Community Association (OCA)",
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/__manifest__.py]:
```suggestion
    "author": "Ecosoft, Odoo Community Association (OCA)",
```

File: hr_expense_advance_clearing/models/hr_expense.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+class HrExpense(models.Model):
+    _inherit = "hr.expense"
+
+    advance = fields.Boolean(string="Employee Advance", default=False)
+
+    @api.constrains("advance")
+    def _check_advance(self):
+        for expense in self.filtered("advance"):
+            emp_advance = self.env.ref(
+                "hr_expense_advance_clearing." "product_emp_advance"
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/models/hr_expense.py]:
```suggestion
                "hr_expense_advance_clearing.product_emp_advance"
```

File: hr_expense_advance_clearing/models/hr_expense.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            self.product_id = self.env.ref(
+                "hr_expense_advance_clearing.product_emp_advance"
+            )
+        else:
+            self.product_id = False
+
+    def _get_account_move_line_values(self):
+        move_line_values_by_expense = super()._get_account_move_line_values()
+        # Only when do the clearing, change cr payable to cr advance
+        emp_advance = self.env.ref("hr_expense_advance_clearing." "product_emp_advance")
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/models/hr_expense.py]:
```suggestion
        emp_advance = self.env.ref("hr_expense_advance_clearing.product_emp_advance")
```

File: hr_expense_advance_clearing/models/hr_expense_sheet.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                and len(sheet.expense_line_ids) == 1
+            )
+        return
+
+    @api.constrains("advance_sheet_id", "expense_line_ids")
+    def _check_advance_expense(self):
+        advance_lines = self.expense_line_ids.filtered("advance")
+        if self.advance_sheet_id and advance_lines:
+            raise ValidationError(
+                _("Advance clearing must not contain any " "advance expense line")
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/models/hr_expense_sheet.py]:
```suggestion
                _("Advance clearing must not contain any advance expense line")
```

File: hr_expense_advance_clearing/models/hr_expense_sheet.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            )
+        if advance_lines and len(self.expense_line_ids) != 1:
+            raise ValidationError(
+                _("Advance must contain only 1 " "advance expense line")
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/models/hr_expense_sheet.py]:
```suggestion
                _("Advance must contain only 1 advance expense line")
```

File: hr_expense_advance_clearing/models/hr_expense_sheet.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            )
+
+    @api.depends("account_move_id.line_ids.amount_residual")
+    def _compute_residual(self):
+        emp_advance = self.env.ref(
+            "hr_expense_advance_clearing." "product_emp_advance", False
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/models/hr_expense_sheet.py]:
```suggestion
            "hr_expense_advance_clearing.product_emp_advance", False
```

File: hr_expense_advance_clearing/models/hr_expense_sheet.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        for sheet in self:
+            rec_lines = sheet.account_move_id.line_ids.filtered(
+                lambda x: x.credit and x.account_id.reconcile and not x.reconciled
+            )
+            sheet.amount_payable = -sum(rec_lines.mapped("amount_residual"))
+
+    def action_sheet_move_create(self):
+        res = super(HrExpenseSheet, self).action_sheet_move_create()
+        # Reconcile advance of this sheet with the advance_sheet
+        emp_advance = self.env.ref("hr_expense_advance_clearing." "product_emp_advance")
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/models/hr_expense_sheet.py]:
```suggestion
        emp_advance = self.env.ref("hr_expense_advance_clearing.product_emp_advance")
```

File: hr_expense_advance_clearing/models/hr_expense_sheet.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                .sudo()
+                .search([("id", "in", move_lines.ids), ("account_id", "=", account_id)])
+            )
+            adv_move_lines.reconcile()
+        return res
+
+    def open_clear_advance(self):
+        self.ensure_one()
+        action = self.env.ref(
+            "hr_expense_advance_clearing." "action_hr_expense_sheet_advance_clearing"
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/models/hr_expense_sheet.py:106]:
```suggestion
            "hr_expense_advance_clearing.action_hr_expense_sheet_advance_clearing"
```

File: hr_expense_advance_clearing/tests/test_hr_expense_advance_clearing.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+                "code": "154000",
+                "name": "Employee Advance",
+                "user_type_id": cls.env.ref(
+                    "account.data_account_type_current_assets"
+                ).id,
+                "reconcile": True,
+            }
+        )
+        cls.emp_advance = cls.env.ref(
+            "hr_expense_advance_clearing." "product_emp_advance"
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/tests/test_hr_expense_advance_clearing.py:48]:
```suggestion
            "hr_expense_advance_clearing.product_emp_advance"
```

File: hr_expense_advance_clearing/wizard/hr_expense_sheet_register_payment.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+    def expense_post_return_advance(self):
+        """ This is opposite operation of expense_post_payment(),
+        it return remaining advance from employee back to company
+        """
+        self.ensure_one()
+        context = dict(self._context or {})
+        active_ids = context.get("active_ids", [])
+        expense_sheet = self.env["hr.expense.sheet"].browse(active_ids)
+        emp_advance = self.env.ref("hr_expense_advance_clearing." "product_emp_advance")
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/wizard/hr_expense_sheet_register_payment.py:36]:
```suggestion
        emp_advance = self.env.ref("hr_expense_advance_clearing.product_emp_advance")
```

**Review by ps-tubtim — COMMENTED** [review]
Functional Test: Ok. I have some comments about code review.

File: hr_expense_advance_clearing/wizard/hr_expense_sheet_register_payment.py (hunk 10 above)
Comment (kittiu) [reply hr_expense_advance_clearing/wizard/hr_expense_sheet_register_payment.py:36]: Done all.

**Comment by kittiu** [comment]
@ps-tubtim ready for review.

File: hr_expense_advance_clearing/views/hr_expense_views.xml [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        </field>
+    </record>
+    <record id="action_hr_expense_sheet_advance_clearing" model="ir.actions.act_window">
+        <field name="name">Advance Clearing</field>
+        <field name="res_model">hr.expense.sheet</field>
+        <field name="view_mode">form</field>
+        <field name="search_view_id" ref="hr_expense_sheet_view_search" />
+        <field
+            name="domain"
+        >[('employee_id.user_id', '=', uid), ('state', '!=', 'cancel')]</field>
```
Comment (ps-tubtim) [inline hr_expense_advance_clearing/views/hr_expense_views.xml]:
```suggestion
        <field name="domain">
            [('employee_id.user_id', '=', uid), ('state', '!=', 'cancel')]
        </field>
```
I think this format is better. What do you think about it?

**Review by ps-tubtim — APPROVED** [review]

File: hr_expense_advance_clearing/views/hr_expense_views.xml (hunk 11 above)
Comment (kittiu) [reply hr_expense_advance_clearing/views/hr_expense_views.xml]: Done.

**Comment by Saran440** [comment]
please rebase commit from OCA Bot
`Squash administrative commits (if any) with the previous commit for reducing commit noise. They are named as "[UPD] README.rst", "[UPD] Update $MODULE.pot", "Update translation files" and similar names, and comes from OCA-git-bot, oca-travis or oca-transbot.`

reference : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#tasks-to-do-in-the-migration

**Review by Saran440 — APPROVED** [review]
Code Reviewed LGTM :+1:

**Pull request merged**